crates/bevy_ecs/src/component.rs

```diff
@@ -2126,42 +2142,80 @@ pub fn component_clone_via_clone<C: Clone + Component>(
+pub fn component_clone_via_reflect(world: &mut DeferredWorld, ctx: &mut ComponentCloneCtx) {
+    // This path will be selected if:
+    // 1. Component has ReflectFromReflect or ReflectDefault registered.
+    // 2. Component has ReflectFromPtr registered.
+    // Otherwise, it will fallback to clone_slow.
+    fn clone_fast(ctx: &mut ComponentCloneCtx) -> Option<()> {
+        let registry = ctx.type_registry()?;
+        let component = {
+            let source_component_reflect = ctx.read_source_component_reflect()?;
             let registry = registry.read();
+            let component_info = ctx.components().get_info(ctx.component_id())?;
+            let type_id = component_info.type_id()?;
+
+            if let Some(reflect_from_reflect) =
+                registry.get_type_data::<bevy_reflect::ReflectFromReflect>(type_id)
+            {
+                reflect_from_reflect.from_reflect(source_component_reflect.as_partial_reflect())
+            } else if let Some(reflect_default) =
+                registry.get_type_data::<bevy_reflect::std_traits::ReflectDefault>(type_id)
+            {
+                let mut component = reflect_default.default();
+                component.apply(source_component_reflect.as_partial_reflect());
+                Some(component)
+            } else {
+                None
+            }?
+        };
+        ctx.write_target_component_reflect(component);
 
-            let component_info = world
-                .components()
-                .get_info(component_id)
-                .expect("Component must be registered");
-            let Some(type_id) = component_info.type_id() else {
-                return;
-            };
-            let Some(reflect_component) =
-                registry.get_type_data::<crate::reflect::ReflectComponent>(type_id)
-            else {
-                return;
-            };
-            let source_component = reflect_component
-                .reflect(world.get_entity(source).expect("Source entity must exist"))
-                .expect("Source entity must have reflected component")
-                .clone_value();
-            let mut target = world
-                .get_entity_mut(target)
-                .expect("Target entity must exist");
-            reflect_component.apply_or_insert(&mut target, &*source_component, &registry);
+        Some(())
+    }
+
+    // This will try to clone component using ReflectComponent.
+    // If the component does not have this type data registered, it will be ignored.
+    fn clone_slow(
+        component_id: ComponentId,
+        source: Entity,
+        target: Entity,
+        registry: &crate::reflect::AppTypeRegistry,
+        world: &mut World,
+    ) -> Option<()> {
+        let registry = registry.read();
+
+        let component_info = world.components().get_info(component_id)?;
+        let type_id = component_info.type_id()?;
```

Comment:
Seems like some of this is duplicated from `clone_fast`. Maybe we could merge the two to avoid double lookups and unnecessary retries (i.e. if `component_info.type_id()?` fails in `clone_fast`, there's no need to try again in `clone_slow`).